### PR TITLE
fixes #676: Updating parsing to match entity spec

### LINKED ISSUE #676: Update code to reflect latest Entity Spec
The xform spec has been updated since the current version of the code. Some of the changes are as follows: 

### Structure:
- [x] Reject form definitions written for future spec versions and only allow ones that start with `2022.1`. More notes [here](https://github.com/getodk/central-backend/issues/649#issuecomment-1310825614).
- [x] The entity system properties (`dataset`, `id`, and `create`) are expected to be found in the attributes of the entity node and and `label` is its own child node.
- [x] Not worrying (much) about entity namespace for now. Tracked at [#680](https://github.com/getodk/central-backend/issues/680).

### Validation:
- [x] ~Property name validation: we already reject `name` and `label` properties but also need to reject names starting with `__`. Maybe want to reject strange characters, too.~ **Now tracked at [#688](https://github.com/getodk/central-backend/issues/688).**
- [x] Dataset name validation: make it case **sensitive** (currently case-insensitive). Notes [here](https://github.com/getodk/central-backend/issues/649#issuecomment-1302139099).
- [x] Dataset name validation: make sure to trim whitespace. Notes [here](https://github.com/getodk/central-backend/issues/649#issuecomment-1301693117).
- [x] ~Maybe disallow `itemsets` as dataset name??~ [#673](https://github.com/getodk/central-backend/issues/673)
- [x] Lowercase the entity uuid (or maybe make it case-insensitive in the database?). Notes [here](https://github.com/getodk/central-backend/pull/576#discussion_r1015861945).
- [x] Only create entity if `create` is `1` or `true`

### PR DESCRIPTION
**Pull request by ktuite**
Closes #676

#### What has been done to verify that this works as intended?

Tests and trying it out end-to-end in central. Here's a full entity registration form in xml:

```xml
<?xml version="1.0"?>
<h:html xmlns="http://www.w3.org/2002/xforms" xmlns:entities="http://www.opendatakit.org/xforms/entities" xmlns:ev="http://www.w3.org/2001/xml-events" xmlns:h="http://www.w3.org/1999/xhtml" xmlns:jr="http://openrosa.org/javarosa" xmlns:odk="http://www.opendatakit.org/xforms" xmlns:orx="http://openrosa.org/xforms" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <h:head>
    <h:title>Trees registration with gps</h:title>
    <model odk:xforms-version="1.0.0" entities:entities-version="2022.1.0">
      <instance>
        <data id="trees_registration_gps" version="2022.11.18.01">
          <location/>
          <species/>
          <meta>
            <instanceID/>
            <entity dataset="trees" id="" create="">
              <label/>
            </entity>
          </meta>
        </data>
      </instance>
      <bind nodeset="/data/location" type="geopoint" entities:saveto="geometry" />
      <bind nodeset="/data/species" type="string" entities:saveto="species" />

      <bind jr:preload="uid" nodeset="/data/meta/instanceID" readonly="true()" type="string"/>

      <bind nodeset="/data/meta/entity/@id" type="string"/>
      <setvalue event="odk-instance-first-load" ref="/data/meta/entity/@id" value="uuid()"/>
      <bind nodeset="/data/meta/entity/@create" type="string" calculate="true()"/>
      <bind nodeset="/data/meta/entity/label" calculate="/data/species"  type="string"/>
    </model>
  </h:head>
  <h:body>
    <input ref="/data/species">
      <label>Species of Tree</label>
    </input>
    <input appearance="placement-map" ref="/data/location">
        <label>Location Position</label>
        <hint>Choose where on a map this is</hint>
    </input>
  </h:body>
</h:html>
```

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Matching the current spec is good!

#### Does this change require updates to the API documentation? If so, please update docs/api.md as part of this PR.

Spec documentation is being updated separately and is what is driving this PR.

#### Before submitting this PR, please make sure you have:

- [x] run `make test-full` and confirmed all checks still pass OR confirm CircleCI build passes

### CODE REVIEW

File: lib/data/dataset.js [hunk 1]
```diff
@@ -44,10 +44,12 @@ const validateDatasetName = (name) => {
 const getDataset = (xml) => {
   const metaNode = findOne(root('html'), node('head'), node('model'), node('instance'), node(), node('meta'));
   return traverseXml(xml, [
+    findOne(root('html'), node('head'), node('model'))(attr('entities-version')),
     metaNode(findOne(root(), node('entity'))(tree())),
     metaNode(findOne(root(), and(node('entity'), hasAttr('dataset')))(attr('dataset')))
-  ]).then(([ entityTag, datasetName ]) => {
-    // check if <entity> tag exists at all
+  ]).then(([ version, entityTag, datasetName ]) => {
+    if (version.isEmpty() || version.get() !== '2022.1.0')
```
Comment (lognaturel) [inline lib/data/dataset.js:51]: We only want to reject if the first two parts don't match! That will let us track changes to the spec that don't require consumer changes. So basically if it doesn't start with 2022.1, reject.

File: lib/data/entity.js [hunk 2]
```diff
@@ -42,9 +42,14 @@ const validateEntity = (entity) => {
   if (matches == null) throw Problem.user.entityViolation({ reason: `ID [${id}] is not a valid UUID.` });
   const uuid = matches[2];
 
+  // If create is not an allowed value of true, don't return entity data
+  if (entity.system.create !== '1' || entity.system.create === 'true')
```
Comment (lognaturel) [inline lib/data/entity.js:46]: Shouldn't this be either `entity.system.create !== '1' && entity.system.create !== 'true'` or `!(entity.system.create === '1' || entity.system.create === 'true')`

**Comment by lognaturel** [comment]
Sorry about pouncing on your draft PR. Can you tell I'm excited? 🥳

File: lib/data/dataset.js [hunk 3]
```diff
@@ -44,12 +44,14 @@ const validateDatasetName = (name) => {
...
+    findOne(root('html'), node('head'), node('model'))(attr('entities-version')),
     metaNode(findOne(root(), node('entity'))(tree())),
     metaNode(findOne(root(), and(node('entity'), hasAttr('dataset')))(attr('dataset')))
-  ]).then(([ entityTag, datasetName ]) => {
-    // check if <entity> tag exists at all
+  ]).then(([ version, entityTag, datasetName ]) => {
     if (entityTag.isEmpty() && datasetName.isEmpty())
       return Option.none();
+    if (version.isEmpty() || !version.get().startsWith('2022.1.'))
+      throw Problem.user.invalidEntityForm({ reason: 'Version is not supported.' });
```
Comment (lognaturel) [inline lib/data/dataset.js:54]: This message will be visible to API users, right? Maybe "Entities specification version not supported?" Could you include the version code from the form?

File: test/unit/data/entity.js [hunk 4]
```diff
@@ -41,6 +41,48 @@ describe('extracting entities from submissions', () => {
...
+
+      it('should create entity if create is "true"', () =>
+        fieldsFor(testData.forms.simpleEntity)
+          .then((fields) => fields.filter((field) => field.propertyName || field.path.indexOf('/meta/entity') === 0))
+          .then((fields) => parseSubmissionXml(fields, testData.instances.simpleEntity.one.replace('create="1"', 'create="true"')))
+          .then((result) => {
+            should.exist(result);
+          }));
+
+      it('should return null if entity if create is false', () =>
```
Comment (lognaturel) [inline test/unit/data/entity.js:61]: Extra `if entity`?

File: test/unit/data/entity.js [hunk 5]
```diff
@@ -41,6 +41,48 @@ describe('extracting entities from submissions', () => {
...
+        fieldsFor(testData.forms.simpleEntity)
+          .then((fields) => fields.filter((field) => field.propertyName || field.path.indexOf('/meta/entity') === 0))
+          .then((fields) => parseSubmissionXml(fields, testData.instances.simpleEntity.one.replace('create="1"', 'create="false"')))
+          .then((result) => {
+            should.not.exist(result);
+          }));
+
+      it('should return null if entity if create is "0"', () =>
```
Comment (lognaturel) [inline test/unit/data/entity.js:69]: Extra `if entity`?

File: test/unit/data/entity.js [hunk 6]
```diff
@@ -41,6 +41,48 @@ describe('extracting entities from submissions', () => {
...
+        fieldsFor(testData.forms.simpleEntity)
+          .then((fields) => fields.filter((field) => field.propertyName || field.path.indexOf('/meta/entity') === 0))
+          .then((fields) => parseSubmissionXml(fields, testData.instances.simpleEntity.one.replace('create="1"', 'create="0"')))
+          .then((result) => {
+            should.not.exist(result);
+          }));
+
+      it('should return null if entity if create is something else weird', () =>
```
Comment (lognaturel) [inline test/unit/data/entity.js:77]: Extra `if entity`?

**Review by lognaturel — APPROVED** [review]
Looks great to me! Couple of test names to fix and maybe an error message that could be slightly improved but otherwise all the pieces seem there and well-tested.

**Review by matthew-white — APPROVED** [review]
I reviewed c2e2961bb0f90927ac5d3a5dfc414cb283a1480e and 4f5a1d9e4dba42d8e46a30853510064a90094a9e, which look great. LGTM!

**Pull request merged**